src/coord/src/coord.rs

```diff
@@ -2758,135 +2764,185 @@ impl Coordinator {
             },
         )?;
 
-        // If this optimizes to a constant expression, we can immediately return the result.
-        let resp = if let MirRelationExpr::Constant { rows, typ: _ } = &*source {
-            let rows = match rows {
+        // We create a dataflow and optimize it, to determine if we can avoid building it.
+        // This can happen if the result optimizes to a constant, or to a `Get` expression
+        // around a maintained arrangement.
+        let typ = source.typ();
+        let key: Vec<MirScalarExpr> = typ
+            .default_key()
+            .iter()
+            .map(|k| MirScalarExpr::Column(*k))
+            .collect();
+        // Two transient allocations. We could reclaim these if we don't use them, potentially.
+        // TODO: reclaim transient identifiers in fast path cases.
+        let view_id = self.allocate_transient_id()?;
+        let index_id = self.allocate_transient_id()?;
+        // The assembled dataflow contains a view and an index of that view.
+        let mut dataflow = DataflowDesc::new(format!("temp-view-{}", view_id));
+        dataflow.set_as_of(Antichain::from_elem(timestamp));
+        self.dataflow_builder()
+            .import_view_into_dataflow(&view_id, &source, &mut dataflow);
+        dataflow.export_index(
+            index_id,
+            IndexDesc {
+                on_id: view_id,
+                keys: key,
+            },
+            typ,
+        );
+        // Finalization optimizes the dataflow as much as possible.
+        let dataflow_plan = self.finalize_dataflow(dataflow);
+
+        // At this point, `dataflow_plan` contains our best optimized dataflow.
+        // We will check the plan to see if there is a fast path to escape full dataflow construction.
+        let mut fast_path = None;
+        // We need to restrict ourselves to settings where the inserted transient view is the first thing
+        // to build (no dependent views). There is likely an index to build as well, but we may not be sure.
+        if dataflow_plan.objects_to_build.len() >= 1
+            && dataflow_plan.objects_to_build[0].id == view_id
+        {
+            match &dataflow_plan.objects_to_build[0].view {
+                // In the case of a constant, we can return the result now.
+                dataflow::Plan::Constant { rows } => {
+                    fast_path = Some(Ok(rows.clone()));
```

Comment:
Is there a convenient way to avoid this clone?

Reply:
I'll check it out. It would need to be allowed to consume the `dataflow`, which we can reason will not be required in this case, but .. which needs a bit of dancing around.

Reply:
Can the dataflow go into the `Err` tuple? Dunno, making stuff up.

Reply:
It can. I'm most of the way through writing its own enum, and will try and stash it there. Might be some borrow checker fights; will report. I think there will be a clone in any case, as the plan extracted from the catalog will have rows in there, so something does need to clone them.